src/WebJobs.Extensions.Http/ClaimsIdentityHelper.cs

```diff
@@ -0,0 +1,113 @@
+
+        // Order is important, as it ensure EasyAuth user identity takes precedence over functions key identity
+        private static readonly string[] PotentialIdentityHeaders = new string[] { "x-ms-client-principal", "x-ms-functions-key-identity" };
+
+        private static readonly string[] EasyAuthDependentHeaders = new string[] { "x-ms-client-principal" };
+
+        /// <summary>
+        /// Adds a serialized ClaimsIdentity object to a specified http header
+        /// </summary>
+        /// <param name="request">The HTTP request to add the header to</param>
+        /// <param name="identity">The ClaimsIdentity to be serialized</param>
+        /// <param name="headerName">The name of the header to add the serialized value to</param>
+        public static void AddIdentityToHttpRequest(HttpRequest request, ClaimsIdentity identity, string headerName)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (headerName == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            ClaimsIdentitySlim identitySlim = ClaimsIdentitySlim.FromClaimsIdentity(identity);
+            using (var stream = new MemoryStream())
+            {
+                Serializer.WriteObject(stream, identitySlim);
+                string encodedValue = Convert.ToBase64String(stream.GetBuffer(), 0, (int)stream.Position);
+                request.Headers.Add(headerName, encodedValue);
+            }
+        }
+
+        internal static ClaimsIdentity FromBase64EncodedJson(string payload)
+        {
+            using (var buffer = new MemoryStream(Convert.FromBase64String(payload)))
+            {
+                ClaimsIdentitySlim slim = (ClaimsIdentitySlim)Serializer.ReadObject(buffer);
+                return slim.ToClaimsIdentity();
+            }
+        }
+
+        internal static ClaimsIdentity[] GetClaimsIdentitiesFromHeaders(IDictionary<string, string> headers, bool isEasyAuthEnabled)
+        {
+            var requestHeaders = new Dictionary<string, string>(headers, StringComparer.OrdinalIgnoreCase);
+
+            string[] eligibleHeaders;
+            if (isEasyAuthEnabled)
+            {
+                eligibleHeaders = PotentialIdentityHeaders;
+            }
+            else
+            {
+                eligibleHeaders = PotentialIdentityHeaders.Except(EasyAuthDependentHeaders).ToArray();
+            }
+
+            ClaimsIdentity[] identities = eligibleHeaders
+                .Where(potentialHeader => requestHeaders.ContainsKey(potentialHeader))
+                .Select(headerKey => ClaimsIdentityHelper.FromBase64EncodedJson(requestHeaders[headerKey]))
+                .ToArray();
+            return identities;
+        }
+
+        internal static bool IsEasyAuthEnabled(INameResolver nameResolver)
```

Comment:
I think this should use IEnvironment to access environment, not INameResolver

Reply:
It looks like IEnvironment is defined in the host repo, not in WebJobs, so there is no way for me to pull it in. Is there some other interface you were thinking of?

---
In reply to: [216387019](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/404#discussion_r216387019) [](ancestors = 216387019)

Reply:
Yeah, I guess I meant IConfiguration. Basically you're just reading an environment variable, so you don't need "name resolver" behavior. Behind the scenes, the default INameResolver is just using IConfiguration anyways.

However, I see other extensions are using INameResolver as you are, so you can leave as is if you want.